### PR TITLE
CUPTI: Add high-level wrappers for the callback API.

### PR DESCRIPTION
**Pull request by maleadt**
As part of that, extend the insufficient privileges error so that it's useful everywhere, fixing https://github.com/JuliaGPU/CUDA.jl/issues/2237.

### CODE REVIEW

File: src/compiler/reflection.jl [hunk 1]
```diff
@@ -64,35 +45,30 @@ function code_sass(io::IO, job::CompilerJob; raw::Bool=false)
...
-            CUPTI.cuptiUnsubscribe(subscriber)
-        end
+    compiled = compile(job)
+    CUPTI.enable!(cfg)
+    try
+        link(job, compiled)
+    finally
+        CUPTI.disable!(cfg)
     end
+
```
Comment (vchuravy) [inline src/compiler/reflection.jl]: I prefer making the lifetime requirement explicit. `CUPTI.enable!` leaks the cfg pointer to C and `disable!` ends the pointer leakage. Or we could document that enable!/disable! need to be explicitly matched and a `CUPTI.enable!(cfg)` on it's own is disallowed.

```suggestion
    GC.@preserve cfg begin
        CUPTI.enable!(cfg)
        try
            link(job, compiled)
        finally
            CUPTI.disable!(cfg)
        end
    end
```

Comment (maleadt) [reply src/compiler/reflection.jl]: Yeah I wasn't sure whether that's needed given Julia's semantics. The `cfg` object needs to be kept alive for the `disable!` call, but I'm not sure what SROA does with that.

Comment (vchuravy) [reply src/compiler/reflection.jl]: I hate reasoning about the semantics of optimizations xD That's why I prefer explicit managing of lifetimes in those situations.

1. `pointer_from_objref` should either be inside a `GC.@preserve` or after a manual escape to a global.
2. SROA should only be allowed to touch mutable objects iff there are no escapes
3. For the higher-level optimizer  `pointer_from_objref` or the corresponding `jl_value_ptr` have no additional semantics
    (I think that means `jl_value_ptr` counts as a momentary escape and thus SROA should not happen in the higher-level)
4. In alloc-opt I think `jl_value_ptr` is an address escape thus preventing heap-to-stack

So I think the code as written now is safe enough, but wouldn't be if someone starts moving `CUPTI.disable!(cfg)` or commenting it out.

Comment (maleadt) [reply src/compiler/reflection.jl]: Exposing `GC.@preserve` is just so ugly. Maybe a macro, or yet another closure?

**Pull request merged**